core/tasks.rs

```diff
@@ -0,0 +1,273 @@
+  pub fn poll_inner(&self, cx: &mut Context) -> Poll<Vec<UnsendTask>> {
+    // Check the flag first -- if it's false we definitely have no tasks
+    if !self
+      .has_tasks
+      .swap(false, std::sync::atomic::Ordering::SeqCst)
+    {
+      self.waker.register(cx.waker());
+      return Poll::Pending;
+    }
+
+    let mut lock = self.tasks.lock().unwrap();
+    let tasks = std::mem::take(lock.deref_mut());
+    if tasks.is_empty() {
+      // Unlikely race lost -- the task submission to the queue and flag are not atomic, so it's
+      // possible we ended up with an extra poll here. This only shows up under Miri, but as it is
+      // possible we do need to handle it.
+      self.waker.register(cx.waker());
+      return Poll::Pending;
+    }
+
+    // SAFETY: we are removing the Send trait as we return the tasks here to prevent
+    // these tasks from accidentally leaking to another thread.
+    let tasks = unsafe { std::mem::transmute(tasks) };
+    Poll::Ready(tasks)
+  }
+
+  fn spawn(&self, task: SendTask) {
+    self.tasks.lock().unwrap().push(task);
+    // TODO(mmastrac): can we use a looser ordering here?
+    self
+      .has_tasks
+      .store(true, std::sync::atomic::Ordering::SeqCst);
+    self.waker.wake();
+  }
+}
+
+/// Allows for submission of v8 tasks on the same thread.
+#[derive(Clone)]
+pub struct V8TaskSpawner {
+  // TODO(mmastrac): can we split the waker into a send and !send one?
+  tasks: Arc<V8TaskSpawnerFactory>,
+  _unsend_marker: PhantomData<*const ()>,
+}
+
+impl V8TaskSpawner {
+  /// Spawn a task that runs within the [`crate::JsRuntime`] event loop from the same thread
+  /// that the runtime is running on. This function is re-entrant-safe and may be called from
+  /// ops, from outside of a [`v8::HandleScope`] in a plain `async`` task, or even from within
+  /// another, previously-spawned task.
+  ///
+  /// The task is handed off to be run the next time the event loop is polled, and there are
+  /// no guarantees as to when this may happen.
+  ///
+  /// # Important Notes
+  ///
+  /// The task shares the same [`v8::HandleScope`] as the core event loop, which means that it
+  /// must maintain the scope in a valid state to avoid corrupting or destroying the runtime.
+  ///
+  /// For example, if the code called by this task can raise an exception, the task must ensure
+  /// that it calls that code within a new [`v8::TryCatch`] to avoid the exception leaking to the
+  /// event loop's [`v8::HandleScope`].
+  pub fn spawn<F>(&self, f: F)
+  where
+    F: FnOnce(&mut v8::HandleScope) + 'static,
+  {
+    // SAFETY: we are transmuting Send into a !Send handle but we can guarantee this object will never
+    // leave the current thread because `V8TaskSpawner` is !Send.
+    let task: Box<dyn FnOnce(&mut v8::HandleScope<'_>)> = Box::new(f);
+    let task: Box<dyn FnOnce(&mut v8::HandleScope<'_>) + Send> =
+      unsafe { std::mem::transmute(task) };
```

Comment:
```suggestion
    let task: Box<dyn FnOnce(&mut v8::HandleScope<'_>)> = Box::new(f);
    // SAFETY: we are transmuting Send into a !Send handle but we guarantee this object will never
    // leave the current thread because `V8TaskSpawner` is !Send.
    let task: Box<dyn FnOnce(&mut v8::HandleScope<'_>) + Send> =
      unsafe { std::mem::transmute(task) };
```

Moves safety comment above `unsafe` usage.